fix: Mitigate GutenbergKit text composition oddities

dcalhoun opened the pull request:
## Description
<!-- Describe the changes, why they are needed, and how they address the issue -->
Fix CMM-353. Fix CMM-199.

Mitigate text composition oddities where editor history was lost and the inline
block inserter erroneously dismissed.

Related: 
* https://github.com/wordpress-mobile/GutenbergKit/pull/134
* https://github.com/wordpress-mobile/GutenbergKit/issues/130
* https://github.com/wordpress-mobile/GutenbergKit/issues/103

## Testing instructions
<!-- Consider listing specific testing steps for helping reviewers evaluate the efficacy of the changes -->

> [!Important]
> A bug remains present for Samsung keyboard users where the inline inserter does not update its filtering until text composition ends—adding a space, tapping a word in the toolbar, etc. 

**1: History stack functions as expected**
See: https://github.com/wordpress-mobile/GutenbergKit/issues/130

**2: Inline inserter does not prematurely dismiss**
See: https://github.com/wordpress-mobile/GutenbergKit/issues/103

**3: Save on exit includes latest typed content for devices using Samsung keyboard with predictive text**
1. Use a Samsung device. 
2. Ensure the Samsung keyboard is used and "predictive text" is enabled in the settings. 
3. Type a few words into a paragraph in a post. 
4. Do not type a space after the last word, confirm the last word via the keyboard toolbar, blur the text input, or take any other action that "confirms" the last word. 
5. Tap the back button to exit the editor. 
6. Verify the last (unconfirmed and/or pending text composition) word in the post was saved. 

**4: Smoke test the Gutenberg Mobile (GBM) and Aztec editors**
For GBM: Disable the experimental editor. 
For Aztec: P9ugOq-4rN-p2

<!--

Consider testing the following *yourself* before requesting a review:

- Compatibility with WordPress.com sites and self-hosted Jetpack sites
- Both portrait and landscape orientations
- Light and dark modes
- Dynamic font sizes: larger, smaller and bold text
- High contrast mode
- Screen reader experience (e.g., Talkback)
- Languages with large words or with letters/accents not frequently used in English
- Right-to-left languages layout
- Multi-tasking: split view and slide over

-->

dcalhoun committed:
fix: Mitigate Android text composition oddities

Forcibly ending text composition led to odd outcomes--e.g., erroneously
dismissing the inline inserter, resetting editor history stack.

It was originally added to ensure the latest text in the editor is
captured before providing it to the host app for persistence. This
appears necessary for Samsung device keyboards only. For the time being,
text may be lost on these devices if the user does not first blur the
text input before saving or exiting the editor.

dcalhoun committed:
fix: End GutenbergKit text composition during close

Ensure the latest text is captured before attempting to persist the post
content.

dcalhoun committed:
refactor: Improve error handling consistency

The rest of the editor methods throw this exception.

dcalhoun committed:
build: Update GutenbergKit version

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dcalhoun committed:
style: Address lint warnings

dcalhoun committed:
style: Address missing null annotations

dcalhoun committed:
refactor: Reuse existing `doFinish` argument

The existing `doFinish` argument is used to managed finishing the
activity after saving. Because we want to complete text composition for
this same context, reusing this argument seems appropriate.

adalpari commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.kt:
```diff
@@ -1967,20 +1967,25 @@ class EditPostActivity : BaseAppCompatActivity(), EditorFragmentActivity, Editor
      * 3. Invokes the listener method parameter
      */
-    private fun updateAndSavePostAsyncOnEditorExit(listener: OnPostUpdatedFromUIListener?) {
+    private fun updateAndSavePostAsyncOnEditorExit(listener: OnPostUpdatedFromUIListener?,
+                                                   isFinishing: Boolean = false) {
```
```suggestion
    private fun updateAndSavePostAsyncOnEditorExit(
        listener: OnPostUpdatedFromUIListener?,
        isFinishing: Boolean = false
    ) {
```

⛏️ Just following a more Kotlint style

adalpari commented:
> 1. Use a Samsung device.

Unfortunately, I don't have any Samsung device to test this step

> For Aztec: P9ugOq-4rN-p2
Not sure what this action means. I assume the Aztec editor is the native one?

nbradbury commented:
@adalpari I have a Samsung device so I'll test that scenario.

nbradbury commented:
@adalpari I tested the Samsung scenario and it worked as expected, so we're good on that step.

dcalhoun commented:
> > For Aztec: P9ugOq-4rN-p2
> Not sure what this action means. I assume the Aztec editor is the native one?

@adalpari [Aztec](https://github.com/wordpress-mobile/AztecEditor-Android) is the mobile app's implementation of WordPress' ["classic" editor](https://wordpress.org/plugins/classic-editor/). It was the primary editor before [Gutenberg Mobile](https://github.com/wordpress-mobile/gutenberg-mobile) was built.

nbradbury reviewed (approved):
@dcalhoun I think we're good to merge this. :shipit:

dcalhoun commented:
@nbradbury thank you. Would you review the sibling https://github.com/wordpress-mobile/GutenbergKit/pull/134 as well please?

dcalhoun committed:
build: Update GutenbergKit version
